### PR TITLE
CLOUDP-260468: Detect what versions are deprecated by a version (foascli)

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes
_Jira ticket:_ [CLOUDP-260468](https://jira.mongodb.org/browse/CLOUDP-260468)

This PR adds the filtering logic to append the deprecated message to the operation description in the event the operation has deprecated endpoints. I tested the changes by implementing e2e tests that compare the path descriptions of each operation in the generated spec with the one in the vbranch for each version.  

Note: the e2e test currently check only that the Path description and all the operations description within the path match the ones in the vbranch (test/data/split folder)

### CODE REVIEW

File: tools/cli/internal/cli/split/split.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ func (o *Opts) Run() error {
 
 	// make a copy of the oas to avoid modifying the original document when applying filters
 	for _, version := range versions {
-		oasCopy, err := duplicateOas(oas)
+		oasCopy, err := duplicateOas(oas) // @Todo this can go outside the loop maybe !?
```
Comment (blva) [inline tools/cli/internal/cli/split/split.go:62]: yeah, I think we need to duplicate one per file we want to generate and operate the filters on it

Comment (andreaangiolillo) [reply tools/cli/internal/cli/split/split.go:62]: The good news is that I compared the v2024-05-30.json created with foascli with the one in production and the descriptions fields of the responses matched

Comment (blva) [reply tools/cli/internal/cli/split/split.go:62]: wohooooo 🚀🚀🚀amazing news

Comment (blva) [reply tools/cli/internal/cli/split/split.go:62]: we can add this comparison as e2e, get a file we generated in the past and compare. I'll write a ticket for the e2e.

File: tools/cli/test/e2e/cli/split_test.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		var o, e bytes.Buffer
+		cmd.Stdout = &o
+		cmd.Stderr = &e
+		require.NoError(t, cmd.Run(), e.String())
+
+		ValidateVersionedSpec(t, NewValidAtlas20230101YAMLSpecPath(t), "output-2023-01-01.yaml")
+		ValidateVersionedSpec(t, NewValidAtlas20230201YAMLSpecPath(t), "output-2023-02-01.yaml")
+		ValidateVersionedSpec(t, NewValidAtlas20231001YAMLSpecPath(t), "output-2023-10-01.yaml")
+		ValidateVersionedSpec(t, NewValidAtlas20231115YAMLSpecPath(t), "output-2023-11-15.yaml")
+		ValidateVersionedSpec(t, NewValidAtlas20240530YAMLSpecPath(t), "output-2024-05-30.yaml")
```
Comment (andreaangiolillo) [inline tools/cli/test/e2e/cli/split_test.go:34]: I downloaded the specs from the vbranch https://evergreen.mongodb.com/task/mms_v20240710_local_openapi_required_OPENAPI_GENERATE_SPECS_0d22d60cb60f16df36e99bff6aff668cad2bb623_24_07_12_13_57_27

File: tools/cli/test/e2e/cli/split_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+			"-o",
+			"output.yaml",
+		)
+
+		var o, e bytes.Buffer
+		cmd.Stdout = &o
+		cmd.Stderr = &e
+		require.NoError(t, cmd.Run(), e.String())
+
+		ValidateVersionedSpec(t, NewValidAtlas20230101YAMLSpecPath(t), "output-2023-01-01.yaml")
```
Comment (blva) [inline tools/cli/test/e2e/cli/split_test.go:30]: cool, should we add outout-2023 to git ignore or do we want to keep the file in the repo?

Comment (andreaangiolillo) [reply tools/cli/test/e2e/cli/split_test.go:30]: We need them in the repo, the e2e tests is using them (loads the specs and compare them)

File: tools/cli/internal/openapi/openapi3.go [hunk 4]
```diff
@@ -36,6 +39,12 @@ func NewOpenAPI3() *OpenAPI3 {
 func (o *OpenAPI3) CreateOpenAPISpecFromPath(path string) (*load.SpecInfo, error) {
 	openapi3.CircularReferenceCounter = o.CircularReferenceCounter
 	o.Loader.IsExternalRefsAllowed = o.IsExternalRefsAllowed
+	dir, err := os.Getwd()
+	if err != nil {
+		fmt.Println("Error:", err)
+	}
+	fmt.Println("dir:", dir)
```
Comment (blva) [inline tools/cli/internal/openapi/openapi3.go:46]: [nit] leftover?

File: tools/cli/internal/openapi/openapi/filter/mock_filter.go
Comment (blva) [inline tools/cli/internal/openapi/openapi/filter/mock_filter.go:1]: nit: move all mocks to a folder called internal/mocks (copying from the CLI https://github.com/mongodb/mongodb-atlas-cli/tree/master/internal/mocks)

File: tools/cli/test/e2e/cli/split_test.go (hunk 3 above)
Comment (blva) [reply tools/cli/test/e2e/cli/split_test.go:30]: I thought the output one should be generated by the test, no?

File: tools/cli/internal/openapi/openapi/filter/mock_filter.go
Comment (andreaangiolillo) [reply tools/cli/internal/openapi/openapi/filter/mock_filter.go:1]: got a good comment from Gus once reminding to keep the mock close to where the original file is

File: tools/cli/internal/openapi/openapi3.go (hunk 4 above)
Comment (andreaangiolillo) [reply tools/cli/internal/openapi/openapi3.go:46]: yes, thanks!

**Review by blva — APPROVED** [review]

File: tools/cli/test/e2e/cli/split_test.go (hunk 3 above)
Comment (andreaangiolillo) [reply tools/cli/test/e2e/cli/split_test.go:30]: yes, `output-2023-01-01.yaml` is the spec generated by the split command that run in the test which is validated against the `NewValidAtlas20230101YAMLSpecPath` which is the versioned spec downloaded from the vbranch and added to the `test/data/split` folder. To be clear, the test doesn't download the spec from the vbranch. I did it manually

Comment (andreaangiolillo) [reply tools/cli/test/e2e/cli/split_test.go:30]: let me know if something is still missing, I will wait for this thread to be resolved before merging. Thanks

**Pull request merged**